rustdoc: edition flag isn't propagated to doc-tests

Given a file:

```rust
#![feature(async_await)]

/// ```rust
/// #![feature(async_await)]
/// fn main() {
///     let _ = async { };
/// }
/// ```
fn main() {
    let _ = async { };
}
```

compiling it with `rustc --edition 2018 temp.rs` works. Attempting to run the doc-test via `rustdoc -Z unstable-options --edition 2018 --test temp.rs` gives:

```
running 1 test
test temp.rs - main (line 3) ... FAILED

failures:

---- temp.rs - main (line 3) stdout ----
error[E0422]: cannot find struct, variant or union type `async` in this scope
 --> temp.rs:6:13
  |
4 |     let _ = async { };
  |             ^^^^^ not found in this scope

thread 'temp.rs - main (line 3)' panicked at 'couldn't compile the test', librustdoc/test.rs:327:17
note: Run with `RUST_BACKTRACE=1` for a backtrace.


failures:
    temp.rs - main (line 3)

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out
```

which is the same error that `rustc --edition 2015 temp.rs` gives.

EDIT: Tested on versions:

```
rustdoc 1.29.0-nightly (960f6046c 2018-07-08)
rustdoc 1.29.0-nightly (64f7de921 2018-07-12)
```

--- a/src/librustdoc/test.rs
+++ b/src/librustdoc/test.rs
@@ -37,7 +37,7 @@ use syntax::codemap::CodeMap;
 use syntax::edition::Edition;
 use syntax::feature_gate::UnstableFeatures;
 use syntax::with_globals;
-use syntax_pos::{BytePos, DUMMY_SP, Pos, Span, FileName};
+use syntax_pos::{BytePos, DUMMY_SP, Pos, Span, FileName, hygiene};
 use errors;
 use errors::emitter::ColorConfig;
 
@@ -561,6 +561,7 @@ impl Collector {
                     rustc_driver::in_rustc_thread(move || with_globals(move || {
                         io::set_panic(panic);
                         io::set_print(print);
+                        hygiene::set_default_edition(edition);
                         run_test(&test,
                                  &cratename,
                                  &filename,

--- a/src/test/rustdoc/edition-flag.rs
+++ b/src/test/rustdoc/edition-flag.rs
@@ -1,0 +1,24 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// compile-flags:--test -Z unstable-options
+// edition:2018
+
+#![feature(async_await)]
+
+/// ```rust
+/// #![feature(async_await)]
+/// fn main() {
+///     let _ = async { };
+/// }
+/// ```
+fn main() {
+    let _ = async { };
+}

